---
title: "OncoMix Vignette"
author: "Daniel PiquÃ©"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{OncoMix Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

OncoMix is an R package for identifying oncogene candidates based off of 2-component Gaussian mixture models. It estimates parameters using EM as implemented in mclust.
This tutorial will demonstrate how to identify oncogene candidates from a set of mRNA sequencing data.



The advent of large, well-curated databases (such as the [genomic data commons](https://gdc.cancer.gov/) and [oncomine](https://www.oncomine.org/)) that contain RNA sequencing data from hundreds of patient tumors has made it possible to identify oncogene candidates based off of mRNA expression alone. A definition for oncogenes has also been proposed based on mutation calls from high-throughput DNA sequencing data in tumors. Specifically, an oncogene is a gene that is recurrently mutated in 20% or more of patient samples (see Vogelstein et al, Science, 2013). Other attempts at identifying oncogenes from gene expression data include COPA, which identifies translocations from microarray data, and mCOPA, a derivative of COPA that is more flexible and that can identify outlier genes. However, no methods have been developed to identify oncogenes from RNA-sequencing data.

```{r}
#devtools::install_github("dpique/oncomix")
#setwd("\\\\data.einstein.yu.edu/home/dpique/dpLabNotebook/bimodality_brca_tcga/oncomix")
#library(devtools)
#install_github("dpique/oncomix")
#install.packages("oncomix")
#vignette("oncomix_vignette")

#remove.packages("oncomix")
library(oncomix)
```

## What do we expect an oncogene to look like?

```{r}
library(ggplot2)
oncomix::oncoMixIdeal()

```

The teal curves represent the probability distribution (PD) over expression values from a single gene obtained from multiple tumors samples. The red-orange curves represent the PD over expression values from the same gene obtained from multiple adjacent normal tissue samples. In other words, there is a *subgroup* of tumors that overexpresses this gene relative to normal tissue.

## Identifying Oncogenes 


```{r}
data(dfNmlIsof, dfTumorIsof, package="oncomix")

dim(dfNmlIsof)
dfNmlIsof[1:5, 1:5] #take a look at the matrix of mRNA expression data from adjacent normal samples

dim(dfTumorIsof)
dfTumorIsof[1:5, 1:5] #take a look at the matrix of mRNA expression data from tumors

mmParams = mixModelParams(dfNmlIsof, dfTumorIsof) #will take a few mins
head(mmParams)
```


## Visualize the output

```{r, fig.width = 7, fig.height= 6.5}
scatterMixPlot(mmParams)
```



##Now, highlight the isoforms with a selectivity index greater than .99

```{r, fig.width = 7, fig.height= 6.5}
scatterMixPlot(mmParams, selIndThresh = .99)
```

##Create a clickable app that shows the identity of each isoform (and also its distribution in a pop-out menu). Create an option to show known oncogenic isoforms.
```{r}
library(shiny)
#runExample("01_hello")
#runExample("02_text")
#runExample("03_reactivity")
```

