---
title: "OncoMix Vignette"
author: "Daniel PiquÃ©"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{OncoMix Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

OncoMix is an R package for identifying oncogene candidates based off of 2-component Gaussian mixture models. It estimates parameters using EM as implemented in mclust.
This tutorial will demonstrate how to identify oncogene candidates from a set of mRNA sequencing data.



The advent of large, well-curated databases (such as the [genomic data commons](https://gdc.cancer.gov/) and [oncomine](https://www.oncomine.org/)) that contain RNA sequencing data from hundreds of patient tumors has made it possible to identify oncogene candidates based off of mRNA expression alone. A definition for oncogenes has also been proposed based on mutation calls from high-throughput DNA sequencing data in tumors. Specifically, an oncogene is a gene that is recurrently mutated in 20% or more of patient samples (see Vogelstein et al, Science, 2013). Other attempts at identifying oncogenes from gene expression data include COPA, which identifies translocations from microarray data, and mCOPA, a derivative of COPA that is more flexible and that can identify outlier genes. However, no methods have been developed to identify oncogenes from RNA-sequencing data.

```{r}
#devtools::install_github("dpique/oncoMix1")
library(oncoMix)
```

## What do we expect an oncogene to look like?

```{r}
library(ggplot2)
oncoMix::oncoMixIdeal()

```

The teal curves represent the probability distribution (PD) over expression values from a single gene obtained from multiple tumors samples. The red-orange curves represent the PD over expression values from the same gene obtained from multiple adjacent normal tissue samples. In other words, there is a *subgroup* of tumors that overexpresses this gene relative to normal tissue.

## Identifying Oncogenes 


```{r}
data(dfTumor, dfNormal, package="oncoMix")


dfNml = as.data.frame(matrix(data = rgamma(n = 150, shape = 2, rate = 2), nrow = 15, ncol = 10))
rownames(dfNml) = paste0("patient.n", 1:nrow(dfNml))
colnames(dfNml) = paste0("gene", 1:ncol(dfNml))
dfTumor = as.data.frame(matrix(data = rgamma(n = 150, shape = 4, rate = 3), nrow = 15, ncol = 10))
rownames(dfTumor) = paste0("patient.t", 1:nrow(dfTumor))
colnames(dfTumor) = paste0("gene", 1:ncol(dfTumor))
mmParams = mixModelParams(dfNml, dfTumor)


```


## Visualize the output

```{r}
#oncoMix::()
```

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
